cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,110 @@
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the quorum leader id");
+        if (quorumInfoFuture == null) {
+            this.quorumInfoFuture = describeMetadataQuorum();
+        }
+        return quorumInfoFuture.map(QuorumInfo::leaderId).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthyWithoutPod(int podId, QuorumInfo info) {
+        int leaderId = info.leaderId();
+        Map<Integer, Long> replicaStates = info.voters().stream().collect(Collectors.toMap(
+                QuorumInfo.ReplicaState::replicaId,
+                replicaState -> replicaState.lastCaughtUpTimestamp().isPresent() ? replicaState.lastCaughtUpTimestamp().getAsLong() : -1));
+        int totalNumOfControllers = replicaStates.size();
+        AtomicInteger numOfCaughtUpControllers = new AtomicInteger();
+        if (leaderId < 0) {
+            LOGGER.warnCr(reconciliation, "No quorum leader is found because the leader id is set to {}", leaderId);
+            return false;
+        }
+        long leaderLastCaughtUpTimestamp = replicaStates.get(leaderId);
+        LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for the leader replica {} is {}", leaderId, leaderLastCaughtUpTimestamp);
+        replicaStates.forEach((replicaId, lastCaughtUpTimestamp) -> {
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
+            } else {
+                if (replicaId == leaderId || (leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp) < controllerQuorumFetchTimeoutMs) {
+                    // skip the controller that we are considering to roll
+                    if (replicaId != podId) {
+                        numOfCaughtUpControllers.getAndIncrement();
+                    }
+                    LOGGER.debugCr(reconciliation, "The controller {} has caught up with the quorum leader", replicaId);
+                    LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for controller {} is {}", replicaId, lastCaughtUpTimestamp);
+                } else {
+                    LOGGER.debugCr(reconciliation, "The controller {} is behind from the leader", replicaId);
+                }
+            }
+        });
+        LOGGER.debugCr(reconciliation, "Out of {} voters, there are {} controllers that have caught up with the quorum leader, not including the controller {}", totalNumOfControllers, numOfCaughtUpControllers, podId);
+        return numOfCaughtUpControllers.get() >= ceil((double) (totalNumOfControllers + 1) / 2);
```

Comment:
How / where do we handle 1 and 2 node controller quorums? The formula does not seem to work for them. (That is not a problem of the formula - 1 or 2 controller nodes cannot be rolled safely. But we still need to be able to roll them I guess  - unless we want to reject such custom resources as invalid)

Reply:
What Tina meant about configuration, is maybe we could make a change here so that by default if there are only one or two controllers configured the KafkaRoller never restarts them (as it would be unsafe). Then we could add some configuration field that allows users to explicitly opt in to letting the KafkaRoller restart their 1 or 2 node controller quorum. That way users can run with a quorum of 1 or 2, and are not caught out by the KafkaRoller restarting the pods unexpectedly

Reply:
I would expect this to work the same way as ZooKeeper does today TBH.

Reply:
For ZooKeeper in Strimzi we just happily roll it right? And the operator adds a warning to the Kafka CR that only having one ZooKeeper node is unsafe?

Reply:
Correct.

Reply:
(If the warning is not present yet, I would suggest to add it in separate PR and just solve the rolling to get this merged as soon as possible)

Reply:
Warning has been added for 1 or 2 nodes case following the ZK behaviour.